spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpOutboundGateway.java

```diff
@@ -172,7 +196,9 @@ protected Object handleRequestMessage(Message<?> requestMessage) {
 			throw new MessageHandlingException(requestMessage, "Interrupted in the [" + this + ']', e);
 		}
 		finally {
-			cleanUp(haveSemaphore, connection, connectionId);
+			if (!isAsync()) {
```

Comment:
Can we obtain `isAsync()` as local variable only once in the begging of method?
Kinda to avoid a race condition when it could be changed in between at runtime...